### PR TITLE
8051627: Invariants about java.net.URI resolve and relativize are wrong

### PR DESCRIPTION
**Pull request by tkiriyama**
The current documentation of URI class describes relationship between resolve and relativize methods as follows.

```
For any two normalized URIs u and v,
    u.relativize(u.resolve(v)).equals(v) and
    u.resolve(u.relativize(v)).equals(v) .
```

However, there are some cases where this relationship is not true. For example, when u and v are "http://a/b" and "c/d", respectively, both u.relativize(u.resolve(v)) and u.resolve(u.relativize(v)) equal to "http://a/c/d" not "c/d". The documentation should be updated to avoid confusion by describing the relationship between resolve nd relativize methods under certain conditions of the URI. Note that these conditions are sufficient but not necessary for the statement.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8051627](https://bugs.openjdk.org/browse/JDK-8051627): Invariants about java.net.URI resolve and relativize are wrong
 * [JDK-8291462](https://bugs.openjdk.org/browse/JDK-8291462): Invariants about java.net.URI resolve and relativize are wrong (**CSR**)

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9671/head:pull/9671` \
`$ git checkout pull/9671`

Update a local copy of the PR: \
`$ git checkout pull/9671` \
`$ git pull https://git.openjdk.org/jdk pull/9671/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9671`

View PR using the GUI difftool: \
`$ git pr show -t 9671`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9671.diff">https://git.openjdk.org/jdk/pull/9671.diff</a>

</details>

### CODE REVIEW

**Comment by tkiriyama** [comment]
Could anyone please reviwe this fix?

File: src/java.base/share/classes/java/net/URI.java [hunk 1]
```diff
@@ -219,12 +219,20 @@
  * {@code demo/b/index.html}
  * </blockquote>
  *
- * <p> <i>Relativization</i>, finally, is the inverse of resolution: For any
- * two normalized URIs <i>u</i> and&nbsp;<i>v</i>,
+ * <p> <i>Relativization</i>, finally, can be regarded as the inverse of resolution.
+ * Let <i>u</i> be any normalized absolute URI ending with a slash character ({@code '/'})
+ * and <i>v</i> be any normalized relative URI not beginning with a period character ({@code '.'})
+ * or slash character ({@code '/'}). Then, the following statement is true.
```
Comment (dfuch) [inline src/java.base/share/classes/java/net/URI.java:225]: A colon ':' may be more appropriate than a period '.' at the end of this sentence:
I'd suggest:
> Then, the following statement is true:

File: src/java.base/share/classes/java/net/URI.java [hunk 2]
```diff
@@ -219,12 +219,20 @@
...
  *
  * <blockquote>
- *   <i>u</i>{@code .relativize(}<i>u</i>{@code .resolve(}<i>v</i>{@code )).equals(}<i>v</i>{@code )}&nbsp;&nbsp;and<br>
- *   <i>u</i>{@code .resolve(}<i>u</i>{@code .relativize(}<i>v</i>{@code )).equals(}<i>v</i>{@code )}&nbsp;&nbsp;.<br>
+ *   <i>u</i>{@code .relativize(}<i>u</i>{@code .resolve(}<i>v</i>{@code )).equals(}<i>v</i>{@code )}
+ * </blockquote>
+ *
+ * Let <i>u</i> be any normalized absolute URI ending with a slash character ({@code '/'})
+ * and <i>v</i> be any normalized absolute URI. Then, the following statement is true.
```
Comment (dfuch) [inline src/java.base/share/classes/java/net/URI.java:232]: Same remark here.

**Review by dfuch — CHANGES_REQUESTED** [review]
I believe you are correct. Could you update the UriTest with some examples and counter-examples to validate this claim?

**Comment by tkiriyama** [comment]
@dfuch 
Thank you for your review and suggestions.

> A colon ':' may be more appropriate than a period '.' at the end of this sentence:
> I'd suggest:

You’re right. I changed to use a colon.

> Could you update the UriTest with some examples and counter-examples to validate this claim?

I added tests in URI/Test.java.  
The tests include testing if the resolve-relativize relationship stand when the two URIs meet the condition of the modified javadoc.  
The tests also include testing if the relationship does not stand when the two specific URIs which meets the condition of the original javadoc are given. These URIs are excluded by the condition of the modiied javadoc.

**Comment by dfuch** [comment]
Hi @tkiriyama sorry for taking so long to get back to you. I had a look at the CSR - and fixed a few missing fields, and added myself as reviewer. Please move its state to "Proposed" to get this moving further.

**Comment by tkiriyama** [comment]
@dfuch 
Thank you for your reply.
I moved CSR's status to "PROPOSED". Please review it.

**Comment by tkiriyama** [comment]
This pull request is waiting for the CSR.

**Comment by tkiriyama** [comment]
@dfuch 
The CSR of this pull request was moved to "CLOSED".   
Could you review this pull request again?

**Review by dfuch — APPROVED** [review]

**Comment by dfuch** [comment]
Done! Thanks for reminding me.

**Comment by dfuch** [comment]
If you need a sponsor please drop a note here after integrating.

**Comment by tkiriyama** [comment]
Thank you for your review. Could you sponsor this pull request, please?

**Pull request closed without merging**